client/app/bundles/course/assessment/submission/pages/SubmissionEditIndex/SubmissionEditForm.jsx

```diff
@@ -114,6 +120,33 @@ const SubmissionEditForm = (props) => {
 
   const methods = useForm({
     defaultValues: initialValues,
+    resolver: async (data) => {
```

Comment:
> * This code is duplicated in `SubmissionEditStepForm`. As the logic here is not simple, you will have a problem in the future when you want to change the code (ie forget to update the code in the other file). Abstract this.

Alright, it's been abstracted out

> * This is a submission level component, but you are coupling it with question-level logic (ie line 127 to 144 below). Can you tell me how to improve this?

The validation for the submission indeed needs to be based on what type of question does the validation take place. For example, to ensure the validity of the answer based on attachment, we should not bother to do that for, let's say, MCQ, MRQ, etc. and only do that for TextResponse and FileUpload.

Though, I understand your concern here that some questions might not even have the isAttachmentRequired attribute. That one, I just realised after trying to abstract this resolver out. Now, in the submission level, we only need to pass on the questions and also attachments into the resolver component to return the validation function for us
 
> * What happens in the future if every single question type has its own validation logic? Put them together here?

I have abstracted out so that when the new validation logic comes, you can put that appropriately inside the proper questionType